### DISCUSSION TITLE
How do you access the access_token in the frontend?

### DISCUSSION THREAD

**Discussion by gooseman1721** [fief-dev/fief#92] [Q&A] [answered]
### The issue

I have a FastAPI endpoint set up like in the tutorial:

```
@app.get("/fief_user/")
async def get_fief_user(
    access_token_info: FiefAccessTokenInfo = Depends(auth.authenticated()),
):
    return access_token_info
```

and I want to call it from my frontend. To do this, I need to call the endpoint with a dictionary that includes an access_token.

I can login to Fief in my frontend. How do I extract the access_token of the current user so I can pass it to the backend?

**Comment by frankie567** [comment, marked as answer]
It depends on how you obtained the access token in your frontend. If you followed the [browser example](https://docs.fief.dev/integrate/javascript/browser/), the token is available through the [`fiefAuth.getTokenInfo` method](https://fief-dev.github.io/fief-js/classes/index.browser.FiefAuth.html#getTokenInfo). It'll return you a [`FiefTokenResponse` object](https://fief-dev.github.io/fief-js/interfaces/index.FiefTokenResponse.html) containing the access token.

Then, you can pass it in the `Authorization` header when making API requests to your backend.

```ts
const tokenResponse = fiefAuth.getTokenInfo();
const accessToken = tokenResponse.access_token;
// Make a request with fetch; but it can be axios or whatever you like to make XHR requests
const response = await fetch(
    'http://localhost:8000/fief/user',
    {
        headers: {
            Authorization: `Bearer ${accessToken}`,
        }
    }
);
data = response.json()
```

**Reply by gooseman1721** [reply]
Thanks for the quick response! Unfortunately, I have encountered another issue. Frontend send a POST to `http://localhost:9000/api/token` with payload

```
grant_type: authorization_code
client_id: <frontend client_id>
code: <code>
redirect_uri: http://127.0.0.1:5173/callback
```

and it returns `{error: "invalid_client"}`.  I double checked and the frontend client_id in the POST is indeed the client_id in the admin dashboard, so I'm not sure what is the problem here.  

EDIT: Perhaps this is due to PKCE? I was under impression that this is handled automatically when using React integration, is that not the case when doing manual fetch like in the example given above?

**Reply by frankie567** [reply]
You're right, it's because the PKCE code is missing. If you use the browser/React integration with the special [`fiefAuth.redirectToLogin`](https://fief-dev.github.io/fief-js/classes/index.browser.FiefAuth.html#redirectToLogin) and [`fiefAuth.authCallback`](https://fief-dev.github.io/fief-js/classes/index.browser.FiefAuth.html#authCallback); then yes, it's handled automatically.

If you make the request yourself to `/token` however, you'll have to take care of passing the code verifier yourself in the request. Two situations then:

1. You want to do everything manually, you'll have to generate the code verifier and code challenge, store them somewhere (e.g. in session storage) before redirecting with the code challenge. Then, after the redirection, call `/token` with the code verifier your retrieved from your storage. This is more or less what is shown here: https://docs.fief.dev/going-further/pkce/#javascript-integration
2. You used  [`fiefAuth.redirectToLogin`](https://fief-dev.github.io/fief-js/classes/index.browser.FiefAuth.html#redirectToLogin) to make the redirection, then our library took care of generating the code verifier and code challenge and store them in session storage. Then, you  retrieve them like this before your request:

```ts
const codeVerifier = fiefAuth.storage.getCodeVerifier();
const params = new URLSearchParams({
    grant_type: 'authorization_code',
    client_id: CLIENT_ID,
    code: AUTHORIZATION_CODE,
    redirect_uri: 'http://127.0.0.1:5173/callback',
    code_verifier: codeVerifier,
});
const response = await fetch(
    `http://localhost:9000/api/token?${params}`, 
);
```

The question is: why do you need to make the `/token` request yourself? Is there something missing you need in the JS client? 😃

**Reply by gooseman1721** [reply]
Perhaps there is something missing. What I want to do is just call an endpoint in my API securely.

I don't want to make /api/token request myself, but it doesn't get called correctly automatically, I don't know why. 

I have a simple login page with:

```
  const fiefAuth = useFiefAuth();
  const isAuthenticated = useFiefIsAuthenticated();
  const userinfo = useFiefUserinfo();

  const login = useCallback(() => {
    fiefAuth.redirectToLogin(
      `${window.location.protocol}//${window.location.host}/callback`
    );
  }, [fiefAuth]);

  const logout = useCallback(() => {
    fiefAuth.logout(`${window.location.protocol}//${window.location.host}`);
  }, [fiefAuth]);

```


and a button that calls login or logout function:

```
              {!isAuthenticated && (
                <Button
                  variant="contained"
                  sx={{ alignSelf: "center" }}
                  onClick={() => login()}
                >
                  Login
                </Button>
              )}
              {isAuthenticated && userinfo && (
                <Button
                  variant="contained"
                  sx={{ alignSelf: "center" }}
                  onClick={() => logout()}
                >
                  Logout {userinfo.email}
                </Button>
              )}
```

After callback, I am logged in and automatically the above mentioned POST to /api/token is made. 

EDIT: It seems that there are actually two POSTs sent to /api/token/. The first contains code_verifier and passes through, the second doesn't and gets rejected.  
 
![chrome_ebROkKpqo3](https://user-images.githubusercontent.com/48566757/194560876-cf1b5de3-73cf-43ab-a065-766c53e052c6.png)

**Reply by frankie567** [reply]
Hmmm 🤔 So it seems that the operation is performed twice, we see that there are two requests to `openid-configuration` and `token`. The first `token` request goes well but the second fails of course.

Could you show the code for the callback page? Maybe there is a weird React re-render causing the operation to be performed twice.

**Reply by gooseman1721** [reply]
It's basically copied from the docs:

FiefCallback.jsx:

```
import { useEffect } from "react";
import { useFiefAuth } from "@fief/fief/react";
import { useNavigate } from "react-router-dom";

export default function FiefCallback() {
  const fiefAuth = useFiefAuth();
  const navigate = useNavigate();

  useEffect(() => {
    fiefAuth
      .authCallback(
        `${window.location.protocol}//${window.location.host}/callback`
      )
      .then(() => {
        navigate("/main");
      });
  }, [fiefAuth, navigate]);

  return <p>Callback</p>;
}
```

My root element looks a little different, as I'm using a different version of react router:

```
ReactDOM.createRoot(document.getElementById("root")).render(
  <React.StrictMode>
    <FiefAuthProvider
      baseURL="http://localhost:9000"
      clientId=<clientId>
    >
      <RouterProvider router={router} />
    </FiefAuthProvider>
  </React.StrictMode>
);
```

**Reply by frankie567** [reply]
So, I was able to reproduce the issue with React 18 with Strict Mode enabled. Apparently, in development mode, [strict mode automatically re-renders components](https://stackoverflow.com/a/72238236) to help discover some bugs or bad patterns.

In my tests though, the first call was working and the session was correctly set, even if the second one was failing. So I would say it shouldn't block you for now.

I'll try to investigate how to avoid this and improve our React integration.

**Reply by gooseman1721** [reply]
Getting rid of Strict Mode does remove the second POST call. I still can't access the access token though. 

`fiefAuth.storage.getCodeVerifier();` doesn't work for me. When I console log fiefAuth.storage, I get:

`{"state":{"userinfo":null,"tokenInfo":null},"sessionStorage":{"test":"session storage test"}}` 

`useFiefUserInfo()` hook, on the other hand, seems to be working, it returns:

`{"at_hash":<at hash>,"aud":[<aud>],"auth_time":<auth time>,"azp":<azp>,"c_hash":<c hash>,"email":<email>,"exp":<exp>,"fields":{"favourite_snail":"snail"},"iat":<iat>,"is_active":true,"is_superuser":false,"is_verified":false,"iss":"https://localhost:9000","sub":"<sub>","tenant_id":<tenant id>} `

Maybe there is some issue with assigning data to storage?

**Reply by gooseman1721** [reply]
Okay, I think I got it. I used `useFiefTokenInfo() ` hook and it got me the access_token of the current user, I was then able to call my API endpoint.

Is that the correct way of doing this?

**Reply by frankie567** [reply]
Yes, that's exactly it!

* `[useFiefUserInfo](https://fief-dev.github.io/fief-js/functions/react.useFiefUserinfo.html)` gives you the information about the user (it's the decoded ID token).
* `[useFiefTokenInfo](https://fief-dev.github.io/fief-js/functions/react.useFiefTokenInfo.html)` give you the result of the `/token` call, including the access token.

It's true that it's not very obvious from the example. I added something about this [in the docs](https://docs.fief.dev/integrate/javascript/react/#call-an-api-with-an-access-token).

---

>  When I console log fiefAuth.storage, I get:
>
> `{"state":{"userinfo":null,"tokenInfo":null},"sessionStorage":{"test":"session storage test"}}`

That's expected if you do this before authentication. After authentication, `userinfo` and `tokenInfo` should be filled with the values described above (`useFiefUserInfo` and `useFiefTokenInfo` actually just read those values in the storage).

`sessionStorage` is here to remember the PKCE code verifier, which is automatically cleared by `authCallback` when the call to `/token` is made.

**Reply by gooseman1721** [reply]
Alright! Good to have that info in the docs now. 

About the storage: 

I have this code in my page:

```
  const fiefAuth = useFiefAuth();
  const isAuthenticated = useFiefIsAuthenticated();
  const tokenResponse = useFiefTokenInfo();

  console.log(isAuthenticated);
  console.log(JSON.stringify(fiefAuth.storage));
  console.log(JSON.stringify(tokenResponse));
```

and the output is:

```
true
{"state":{"userinfo":null,"tokenInfo":null},"sessionStorage":{}}
{"access_token":<access token>,"id_token":<id token>,"token_type":"bearer","expires_in":3600}
```

So I guess that's some different storage the hooks are accessing?
